ear-production-suite-plugins/lib/include/communication/metadata_thread.hpp

```diff
@@ -0,0 +1,75 @@
+//
+// Created by Richard Bailey on 06/04/2022.
+//
+
+#ifndef EAR_PRODUCTION_SUITE_METADATA_THREAD_HPP
+#define EAR_PRODUCTION_SUITE_METADATA_THREAD_HPP
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <vector>
+#include <functional>
+#include <atomic>
+
+namespace ear::plugin {
+class MetadataThread {
+public:
+    using MessageQueue = std::vector<std::function<void()>>;
+    MetadataThread() : thread_([this](){updateLoop();}) {
+    }
+    ~MetadataThread() {
+        stop();
+    }
+
+    void post(std::function<void()> message) {
+        if(run_.load()) {
+            {
+                std::lock_guard<std::mutex> lock(mutex_);
+                messages_.push_back(std::move(message));
+            }
+            condition_.notify_one();
+        }
+    }
+
+
+private:
+    MessageQueue getMessages() {
+        MessageQueue messages;
+        std::unique_lock<std::mutex> lock(mutex_);
+        condition_.wait(lock, [this](){return !messages_.empty() || !run_.load();});
+        messages = messages_;
+        messages_.clear();
+        lock.unlock();
+        return messages;
+    }
+
+    void updateLoop() {
+        {
+            std::lock_guard<std::mutex> lock(mutex_);
+            messages_.reserve(INITIAL_QUEUE_CAPACITY);
+        }
+        while(run_.load()) {
+            auto messages = getMessages();
+            for(auto const& message : messages) {
+                message();
```

Comment:
Do we need to clear messages we've run? it looks like messages in the queue will get run over and over until the messages queue is completely replaced by newly awaiting messages